src/lava/lib/optimization/solvers/generic/solver.py

```diff
@@ -450,20 +482,56 @@ def _prepare_profiler(self, config: SolverConfig, run_cfg) -> None:
             self._profiler = None
 
     def _get_results(self, config: SolverConfig):
-        best_cost, idx = self.solver_process.optimum.get()
-        best_cost = SolutionReadoutPyModel.decode_cost(best_cost)
-        best_state = self._get_best_state(config, idx)
-        best_timestep = self.solver_process.solution_step.aliased_var.get() - 2
-        return best_state, int(best_cost), int(best_timestep)
+        idx = 0
+        if self.solver_process.is_discrete:
+            best_cost, idx = self.solver_process.optimum.get()
+            best_cost = SolutionReadoutPyModel.decode_cost(best_cost)
+            best_timestep = (
+                self.solver_process.solution_step.aliased_var.get() - 2
+            )
+        best_state = self._get_best_state(config, int(idx))
+
+        if self.solver_process.is_discrete:
+            return best_state, int(best_cost), int(best_timestep)
+        else:
+            return best_state, None, None
 
     def _get_best_state(self, config: SolverConfig, idx: int):
+        if self._is_problem_discrete():
+            discrete_values = self._get_and_decode_discrete_vars(config, idx)
+            return discrete_values
+
+        if self._is_problem_continuous():
+            continuous_values = self._get_and_decode_continuous_vars(idx)
+            return continuous_values
+
+    def _is_problem_discrete(self):
+        return (
+            hasattr(self.problem.variables, "discrete")
+            and self.problem.variables.discrete.num_variables is not None
+        )
+
+    def _get_and_decode_discrete_vars(self, config: SolverConfig, idx: int):
         if isinstance(config.hyperparameters, list):
-            idx = int(idx)
             raw_solution = np.asarray(
-                self.solver_process.finders[idx].variables_assignment.get()
-            ).astype(np.int32)
+                self.solver_process.finders[idx]
+                .variables_assignment.get()
+                .astype(np.int32)
+            )
             raw_solution &= 0x3F
             return raw_solution.astype(np.int8) >> 5

```

Comment:
Is this actually working for non-binary discrete value?

Reply:
It works for now. Non-binary discrete values are not in the library. Refactor when the time arises.